### PR TITLE
Introduce explicit ordering for logic rules

### PR DESCRIPTION
**Pull request by sergei-maertens**
Related to #1546 (we need to do this before we can refactor logic in other ways)

- [x] Add explicit order field to logic rule
- [x] Set up data migration to convert existing rules to an explicit order
- [x] Update UI to display logic rules in a single list (merge simple/advanced logic)
- [x] Update UI to allow re-ordering rules & ensure it persists to the backend

Screencast demonstrating the behaviour: https://watch.screencastify.com/v/EoTOgpua3ly4onjKsURz

### CODE REVIEW

File: src/openforms/forms/migrations/0029_make_logic_order_explicit.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    1. Simple rules: introspect the trigger and check which key it is for
+        1. Look up the key in the form components
+        2. Derive which form step the key belongs to
+        3. Order based on form step order
+        4. Within a form step, order on component hierarchy, depth first (i.e. we fetch
+           child components before stepping to siblings.
+        5. If still ambiguous, order by PK
+    2. Simple rules that failed introspection
+    3. Advanced rules: order by PK
```
Comment (sergei-maertens) [inline src/openforms/forms/migrations/0029_make_logic_order_explicit.py:39]: Ewen remarked that the `find_component_key` construct is smart enough to be able to deeply inspect advanced logic rules as well, so we could leverage this to improve the ordering of those. To decide if we want to do this or not!

Comment (sergei-maertens) [reply src/openforms/forms/migrations/0029_make_logic_order_explicit.py:39]: Taking a look at this now

File: src/openforms/forms/api/serializers/logic/form_logic.py [hunk 2]
```diff
@@ -49,11 +54,51 @@ class FormLogicSerializer(FormLogicBaseSerializer):
...
+        as well. Parallel requests affect each other and can lead to unexpected
+        calculated order values.
+
+        To guard against this, we wrap the save operation in an atomic transaction,
+        but this in turn leads to deadlocks because multiple parallel requests are
+        waiting for each other. To mitigate THAT, we lock all the form logic rules for
+        the involved form.
+        """
+        # taken from drf BaseSerializer.save
+        validated_data = validated_data = {**self.validated_data, **kwargs}
```
Comment (SilviaAmAm) [inline src/openforms/forms/api/serializers/logic/form_logic.py:94]:
```suggestion
        validated_data = {**self.validated_data, **kwargs}
```

File: src/openforms/js/components/admin/form_design/form-creation-form.js [hunk 3]
```diff
@@ -524,13 +577,21 @@ function reducer(draft, action) {
...
-      const ruleUuid = draft.logicRules[index].uuid;
+      const {uuid: ruleUuid, order: ruleOrder} = draft.logicRules[index];
       draft.logicRulesToDelete.push(ruleUuid);
 
       // delete object from state
       const updatedRules = [...draft.logicRules];
       updatedRules.splice(index, 1);
       draft.logicRules = updatedRules;
+
+      // update the rules above this rule
```
Comment (SilviaAmAm) [inline src/openforms/js/components/admin/form_design/form-creation-form.js:588]: Maybe a bit clearer?
```suggestion
      // update the rules after this rule
```

File: src/openforms/js/components/admin/forms/JsonWidget.js [hunk 4]
```diff
@@ -5,13 +5,18 @@ import {useIntl} from 'react-intl';
 
 import {TextArea} from './Inputs';
 
+// dump JSON in a readable form
+const jsonFormat = value => {
+  return JSON.stringify(value, null, 2);
```
Comment (SilviaAmAm) [inline src/openforms/js/components/admin/forms/JsonWidget.js:10]: Nice that it is formatted now :slightly_smiling_face:

**Review by SilviaAmAm — APPROVED** [review]
Looks good. Just a couple of questions/minor remarks

**Comment by SilviaAmAm** [comment]
I forgot to mention: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/submissions/form_logic.py#L90

This ordering should now use the new order field

File: src/openforms/forms/models/form.py [hunk 5]
```diff
@@ -382,9 +382,7 @@ def get_keys_for_email_confirmation(self) -> List[str]:
         return list(return_keys)
 
     def iter_components(self, recursive=True):
-        for form_step in self.formstep_set.order_by("order").select_related(
-            "form_definition"
-        ):
+        for form_step in self.formstep_set.select_related("form_definition"):
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/forms/models/form.py:385]: django-ordered-model sets the ordering on the `FormStep.Meta`! :)

File: src/openforms/js/components/admin/form_design/form-creation-form.js (hunk 3 above)
Comment (sergei-maertens) [reply src/openforms/js/components/admin/form_design/form-creation-form.js:588]: I picked the same terminology as used in django-ordered-model for consistency - after/before depends a bit if you order ASC or DESC. This makes more sense though, so I'll change it.

File: src/openforms/js/components/admin/forms/JsonWidget.js (hunk 4 above)
Comment (sergei-maertens) [reply src/openforms/js/components/admin/forms/JsonWidget.js:10]: it's still not great but at least I can read it a bit now :P

File: src/openforms/forms/models/form.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openforms/forms/models/form.py:387]: I added a comment explaining this

**Review by SilviaAmAm — APPROVED** [review]

**Pull request merged**